function.dd

```diff
@@ -121,14 +121,64 @@ int foo();
 
 $(H4 $(LNAME2 pure-functions, Pure Functions))
 
-        $(P Pure functions are functions which cannot access global or static, mutable
-        state save through their arguments. This can enable optimizations based on the fact
-        that a pure function is guaranteed to mutate nothing which isn't passed to it,
+    $(P Pure functions are functions that cannot directly access global or static
+        mutable state. It guarantees that a pure function call won't modify any
+        invisible places from the caller side.
+    )
+
+    $(P Unlike other functional programming languages, D's pure
+        functions allow modification of caller side state through their mutable
+        parameters.
+    )
+
+        ---
+        pure int foo(int[] arr) { arr[] += 1; return arr.length; }
+        int[] a = [1, 2, 3];
+        foo(a);
+        assert(a == [2, 3, 4]);
+        ---
+
+    $(P The minimum guarantee of "pure" is called "weak purity". It's designed to
+        increase modularity of functions.
+        In particular, weak pure functions are good reusable building blocks
+        for strongly pure functions.
```

Comment:
This definition is imprecise. We should specify how a function is weakly impure:

$(P A pure function accepting parameters with mutable indirections offers what's called "weak purity" because it can change program state transitively through its arguments. A pure function that has no parameter with mutable indirections is called "strongly pure" and fulfills the purity definition in traditional functional languages. Weakly pure functions are useful as reusable building blocks for strongly pure functions.)